MRN-Code/coinstac
# Allow Multiple Co-Variates

### Problem

We’re locked to a single co-variate, the `isControl` tag.
### Solution
- Re-work metadata CSV parsing
- Add a new computation input type for covariates
- Add UI for:
  - Attaching required covariates to a consortium
  - Mapping metadata CSVs to required covariates
### Testing
1. Clone [MRN-Code/decentralized-laplacian-ridge-regression](https://github.com/MRN-Code/decentralized-laplacian-ridge-regression) to your machine.
2. Check out the _feature/multiple-covariates_ branch in both _coinstac_ and _decentralized-laplacian-ridge-regression_. Run `npm install` and `npm run bootstrap` respectively to get the freshest dependencies.
3. Create a lovely bash script:
   
   ``` bash
   #!/bin/bash
   
   ALGO="laplacian-noise-ridge-regression@3.0.0"
   COMP_DIR="$(node -e "console.log(require('os').tmpDir() + '/coinstac-server-core/computations')" | tr -d '\r')"
   
   # Change to point to local COINSTAC repo:
   COINSTAC_DIR="$HOME/coinstac"
   
   # Change to point to your local decentralized-laplacian repo:
   TO_LINK_DIR="$HOME/Sites/decentralized-laplacian-ridge-regression"
   
   #1. Start CouchDB
   couchdb -b
   
   #2. Clean the database for a fresh start
   cd $COINSTAC_DIR/packages/coinstac-server-core
   npm run clean:db
   bin/coinstac-server-core --nohang
   
   #3. Clean clients' local databases
   cd $COINSTAC_DIR/packages/coinstac-ui
   npm run clean-db
   
   #4. Link computation
   echo "Stubbing computation in ~/.coinstac/computations"
   mkdir -p ~/.coinstac/computations
   cd ~/.coinstac/computations
   rm -rf $ALGO
   mkdir $ALGO
   echo "module.exports = require('$TO_LINK_DIR')" > $ALGO/index.js
   
   echo "Stubbing computation $COMP_DIR"
   mkdir -p $COMP_DIR
   cd $COMP_DIR
   rm -rf $ALGO
   mkdir $ALGO
   echo "module.exports = require('$TO_LINK_DIR')" > $ALGO/index.js
   ```
   
   …then run it.
4. Fire ‘er up:
   1. Ensure `NODE_ENV=development` in all your shells/subshells
   2. Boot the [mock auth server](https://github.com/swashcap/mock-steelpenny)
   3. Boot the coinstac-server:
      
      ``` shell
      packages/coinstac-server-core/bin/coinstac-server-core --loglevel verbose
      ```
   4. Boot Webpack:
      
      ``` shell
      packages/coinstac-ui/node_modules/.bin/electron packages/coinstac-ui/
      ```
   5. Boot two clients (in two windows):
      
      ``` shell
      packages/coinstac-ui/node_modules/.bin/electron packages/coinstac-ui/
      ```
5. Create a consortium with one client.
   
   ![covariates](https://cloud.githubusercontent.com/assets/1858316/18851697/fd6a99be-83f1-11e6-9831-3a261545a5ac.jpg)
   
   Make sure you select the “3.0.0” algorithm, then add a couple covariates.
6. Join the consortium with the other client.
7. Set up projects on both clients.
   
   ![mapping](https://cloud.githubusercontent.com/assets/1858316/18851740/19049972-83f2-11e6-822a-1a6f0f15300b.jpg)
   
   Map your metadata columns to covariates.
8. Click “Run Computation” in the first client
9. Observe success!
### Related

MRN-Code/decentralized-laplacian-ridge-regression#7
